File: multiqc/plots/echarts/scatter.py
"""
ECharts option builder for `PlotType.SCATTER` (`multiqc/plots/scatter.py`).

Series strategy: when no point carries a `group` field, ONE `{"type": "scatter"}` series
per dataset, with per-item styling (`itemStyle.color`, `symbolSize`), rather than one
series per point. The default template's Plotly `buildTraces()`
(`templates/default/src/js/plots/scatter.js`) builds one Plotly trace per point so every
point can carry its own color/size/legend entry independently; ECharts achieves the same
per-point independence with per-item `data` entries on a single series, which also keeps
`mark_count` (point count) the metric that decides the canvas-vs-svg renderer meaningful:
thousands of series would defeat that threshold's purpose, one series with thousands of
data items does not.

When points DO carry a `group` field (e.g. somalier's relatedness plot), that shortcut
would lose the per-group legend/toggle behaviour that is the entire point of grouping, so
`series()` instead builds ONE series PER GROUP (see `_grouped_series`), ordered per
`pconfig.groups`, mirroring Plotly's `legendgroup`/`inLegend` handling: ECharts gives each
series its own legend entry natively, and clicking it toggles that whole series. This is
still cheap because real grouped datasets have few groups (a handful of relatedness
categories, not thousands of points), so the canvas-threshold rationale above is
unaffected.
"""

from typing import TYPE_CHECKING, Any, Dict, List, Optional, Tuple

from multiqc.plots.echarts.converter import bands_and_lines, convert_layout, trailing_bands_lines_series, zoom_option
from multiqc.plots.scatter import Dataset, PointT, ScatterConfig
from multiqc.utils.mqc_colour import color_to_rgb_string

if TYPE_CHECKING:
    from multiqc.plots.plot import Plot

_DEFAULT_COLOR = "rgba(124, 181, 236, .5)"
_DEFAULT_SIZE = 10

# Plotly marker symbol names (as used by `pconfig.marker_symbol`/`point["marker_symbol"]`)
# mapped to their closest ECharts built-in symbol. ECharts' built-in set is much smaller
# than Plotly's (no per-symbol open/dot variants, no exotic shapes), so this only covers
# the common names with a clean visual equivalent; anything else falls back to "circle",
# same as the interactive JS side (`scatter.js`'s `echartsSymbol()`).
_SYMBOL_MAP: Dict[str, str] = {
    "circle": "circle",
    "square": "rect",
    "diamond": "diamond",
    "triangle": "triangle",
    "triangle-up": "triangle",
}


def _echarts_symbol(marker_symbol: Optional[str]) -> str:
    if marker_symbol is None:
        return "circle"
    return _SYMBOL_MAP.get(marker_symbol, "circle")


def layout_option(plot: "Plot[Any, Any]", dataset: Dataset) -> Dict[str, Any]:
    """
    Full ECharts option skeleton for one scatter-plot dataset, minus `series` (a single
    series built by the JS renderer from the raw points + toolbox state).
    """
    # Scatter has two meaningful value axes (no category axis), so both get Plotly-style
    # data-fitted autorange instead of ECharts' forced-0 default; see
    # `converter._convert_axis`.
    option = convert_layout(plot.layout_ir, dataset.layout, scale_x=True, scale_y=True)

    option["tooltip"]["trigger"] = "item"
    # Plotly-style click+drag box-zoom on both axes (POLISH.md #17): both axes carry
    # real numeric data, see `zoom_option`.
    option.update(zoom_option(x=True, y=True))

    # Scatter has two meaningful value axes (no category axis), so both dimensions
    # apply, same as line.
    bands_lines = bands_and_lines(plot.pconfig, dataset.dconfig)
    if bands_lines:
        option["_mqc"] = {"bandsLines": bands_lines}

    return option


def _point_item(point: PointT, default_color: str, default_size: Any) -> Dict[str, Any]:
    item: Dict[str, Any] = {
        "value": [point["x"], point["y"]],
        "name": point["name"],
        "symbolSize": point.get("marker_size", default_size),
        "symbol": _echarts_symbol(str(point["marker_symbol"]) if "marker_symbol" in point else None),
        "itemStyle": {"color": color_to_rgb_string(str(point.get("color", default_color)))},
    }
    if "opacity" in point:
        item["itemStyle"]["opacity"] = point["opacity"]
    if "marker_line_width" in point:
        item["itemStyle"]["borderWidth"] = point["marker_line_width"]
    return item


def _grouped_series(
    points: List[PointT], pconfig: ScatterConfig, default_color: str, default_size: Any
) -> List[Dict[str, Any]]:
    """
    One `{"type": "scatter"}` series per group, `name` = group name, so ECharts' legend
    shows one entry per group and toggling it hides/shows the whole group (ECharts' native
    per-series legend behaviour standing in for Plotly's `legendgroup`).

    Points are bucketed by `point["group"]`; a point with no group falls back to its own
    `point["name"]` as the bucket key, the same fallback the interactive JS
    (`scatter.js`'s `_buildGroupedSeries`) and the Plotly template's `buildTraces()`
    `inLegend` key use, so a point with no group still gets its own legend entry rather
    than being silently dropped from one.

    Buckets are ordered by first-seen point, then re-sorted by `pconfig.groups` when set
    (groups not listed there go last, keeping their relative first-seen order), mirroring
    the sort in the default template's Plotly `buildTraces()`.
    """
    order: List[str] = []
    buckets: Dict[str, List[PointT]] = {}
    for point in points:
        key = str(point.get("group") or point["name"])
        if key not in buckets:
            buckets[key] = []
            order.append(key)
        buckets[key].append(point)

    if pconfig.groups:
        group_order = pconfig.groups
        order.sort(key=lambda key: group_order.index(key) if key in group_order else len(group_order))

    return [
        {
            "type": "scatter",
            "name": key,
            "data": [_point_item(point, default_color, default_size) for point in buckets[key]],
        }
        for key in order
    ]


def series(dataset: Dataset, pconfig: ScatterConfig, is_pct: bool) -> List[Dict[str, Any]]:
    """
    Scatter series for the SSR/get_option (non-toolbox) path; the interactive path is
    `EchartsScatterPlot.buildSeries()` (`templates/default/src/js/plots-echarts/scatter.js`).

    One `{"type": "scatter"}` series holding every point, each with its own per-item
    style, UNLESS any point carries a `group` field, in which case one series per group
    (see `_grouped_series`) so the legend and click-to-toggle mirror the Plotly template.

    `is_pct` is accepted for dispatch-signature parity with `bar.series`/`line.series`;
    scatter plots never enable the percentage switch (`Plot.initialize` only sets
    `add_pct_tab` for `PlotType.BAR`), so it is unused here.
    """
    marker = dataset.trace_params.get("marker", {})
    default_color = marker.get("color", _DEFAULT_COLOR)
    default_size = marker.get("size", _DEFAULT_SIZE)

    has_groups = any(point.get("group") for point in dataset.points)
    if has_groups:
        result: List[Dict[str, Any]] = _grouped_series(dataset.points, pconfig, default_color, default_size)
    else:
        result = [
            {
                "type": "scatter",
                "name": dataset.label,
                "data": [_point_item(point, default_color, default_size) for point in dataset.points],
            }
        ]

    trailing = trailing_bands_lines_series(pconfig, dataset.dconfig)
    if trailing:
        result.append(trailing)

    return result


def axis_data(dataset: Dataset, pconfig: ScatterConfig) -> Optional[List[Tuple[str, List[str]]]]:
    """
    Always `None`: scatter plots use value (continuous) x/y axes, not a sample-name
    category axis, so there is no axis `data` array for the toolbox to fill in.
    """
    return None


def mark_count(dataset: Dataset) -> int:
    """One mark per point; this is the type most likely to trip the canvas threshold."""
    return len(dataset.points)

File: multiqc/templates/default/src/js/plots-echarts/scatter.js
// Plotly marker symbol names (as used by point.marker_symbol) mapped to their closest
// ECharts built-in symbol. ECharts' built-in set is much smaller than Plotly's (no
// per-symbol open/dot variants, no exotic shapes), so this only covers the common names
// with a clean visual equivalent; anything else falls back to "circle", same as the
// Python side (multiqc/plots/echarts/scatter.py's `_echarts_symbol()`).
const PLOTLY_TO_ECHARTS_SYMBOL = {
  circle: "circle",
  square: "rect",
  diamond: "diamond",
  triangle: "triangle",
  "triangle-up": "triangle",
};

function echartsSymbol(plotlySymbol) {
  if (!plotlySymbol) return "circle";
  return PLOTLY_TO_ECHARTS_SYMBOL[plotlySymbol] ?? "circle";
}

// ECharts scatter plot: extends the default template's ScatterPlot (imported just before
// this file in main-js.js) purely for its prepData()/exportData()/formatDatasetForAiPrompt()/
// activeDatasetSize(), which are engine-neutral (they only read/filter raw dataset fields
// through applyToolboxSettings()). window.ScatterPlot itself extends window.Plot, which by
// this point in the import order is OUR echarts Plot class (from echarts-plotting.js), not
// Plotly's.
//
// Series strategy: when no point carries a `group` field, ONE `{type: "scatter"}` series
// per dataset with per-item styling, not one series per point (see
// multiqc/plots/echarts/scatter.py docstring for why: it keeps a large point cloud a
// single series, which is what makes the canvas-vs-svg renderer threshold meaningful).
// When points DO carry a `group` field, one series PER GROUP instead (see
// buildGroupedSeries), so ECharts' native per-series legend gives Plotly's
// legendgroup/inLegend behaviour: one legend entry per group, click-to-toggle the whole
// group. Real grouped datasets have few groups, so this doesn't defeat the threshold.
class EchartsScatterPlot extends window.ScatterPlot {
  buildSeries() {
    let dataset = this.datasets[this.activeDatasetIdx];
    let marker = dataset["trace_params"]?.["marker"] || {};
    let defaultColor = marker.color ?? "rgba(124, 181, 236, .5)";
    let defaultSize = marker.size ?? 10;

    let [samples, points] = this.prepData();
    // prepData() (inherited from default ScatterPlot) leaves a hole (undefined) for
    // every point the toolbox hides instead of removing it. Compute the "N samples
    // hidden" banner from the hole count before dropping them, same pattern as bar.js.
    this._updateHiddenSamplesWarning(samples.length, points.filter(Boolean).length);
    points = points.filter(Boolean);

    if (points.length === 0 || samples.length === 0) return [];

    // Reorder so highlighted points draw on top, same as default plots/scatter.js.
    let highlighted = points.filter((p) => p.highlight);
    let nonHighlighted = points.filter((p) => !p.highlight);
    points = nonHighlighted.concat(highlighted);

    // Sort points by group order if groups are provided in config, same as default
    // plots/scatter.js buildTraces(). This also determines the group series order below
    // (first-seen order of a now-group-sorted point list matches pconfig.groups).
    if (this.pconfig.groups) {
      points.sort((a, b) => {
        let aIndex = this.pconfig.groups.indexOf(a.group);
        let bIndex = this.pconfig.groups.indexOf(b.group);
        if (aIndex === -1) aIndex = this.pconfig.groups.length;
        if (bIndex === -1) bIndex = this.pconfig.groups.length;
        return aIndex - bIndex;
      });
    }

    // Highlight rule keyed off the GLOBAL flag (not just this plot's matches), matching
    // the dim/grey convention landed for other plot types: once any highlight is active
    // anywhere, every non-matching point greys out.
    let anyHighlighted = window.mqc_highlight_f_texts.length > 0;
    let buildItem = (point) => {
      let color = anyHighlighted ? (point.highlight ?? "#cccccc") : (point.color ?? defaultColor);
      let itemStyle = { color: color };
      if (point.opacity !== undefined) itemStyle.opacity = point.opacity;
      if (point.marker_line_width !== undefined) itemStyle.borderWidth = point.marker_line_width;

      return {
        value: [point.x, point.y],
        name: point.name,
        symbolSize: point.marker_size ?? defaultSize,
        symbol: echartsSymbol(point.marker_symbol),
        itemStyle: itemStyle,
      };
    };

    let hasGroups = points.some((point) => point.group);
    let series = hasGroups
      ? this.buildGroupedSeries(points, buildItem)
      : [
          {
            type: "scatter",
            name: dataset["label"] ?? "",
            data: points.map(buildItem),
          },
        ];

    series.push(...this.bandsLinesSeries());
    return series;
  }

  // One series per group, `name` = group name, so ECharts' legend shows one entry per
  // group and toggling it hides/shows the whole group. Points are bucketed by
  // point.group; a point with no group falls back to its own point.name as the bucket
  // key, the same fallback the default template's Plotly buildTraces() inLegend key
  // uses, so a point with no group still gets its own legend entry rather than being
  // silently folded into another one. `points` is expected to already be sorted by
  // pconfig.groups (buildSeries() does this above), so the first-seen bucket order here
  // already matches.
  buildGroupedSeries(points, buildItem) {
    let order = [];
    let buckets = new Map();
    points.forEach((point) => {
      let key = point.group ?? point.name;
      if (!buckets.has(key)) {
        buckets.set(key, []);
        order.push(key);
      }
      buckets.get(key).push(point);
    });

    return order.map((key) => ({
      type: "scatter",
      name: key,
      data: buckets.get(key).map(buildItem),
    }));
  }

  // See EchartsBarPlot._updateHiddenSamplesWarning (plots-echarts/bar.js) for the full
  // rationale; copied here (kept minimal) since ScatterPlot and BarPlot don't share a
  // common ECharts base beyond window.Plot.
  _updateHiddenSamplesWarning(total, visible) {
    let groupDiv = $("#" + this.anchor).closest(".mqc_hcplot_plotgroup");
    groupDiv.parent().find(".samples-hidden-warning").remove();

    if (visible === 0 && total > 0) {
      groupDiv.hide();
      return;
    }
    groupDiv.show();

    let nHidden = total - visible;
    if (nHidden > 0) {
      const alert = `
      <div class="samples-hidden-warning alert alert-warning">
        ⚠ <strong>Warning:</strong> ${nHidden} samples hidden.
        <a href="#mqc_hidesamples" class="alert-link" onclick="mqc_toolbox_openclose('#mqc_hidesamples', true); return false;">See toolbox.</a>
      </div>`;
      groupDiv.before(alert);
    }
  }

  // Tooltip formatter can't live in the JSON-safe serialized skeleton, so it's attached
  // here to the fully-assembled option (see Plot.applyOptionOverrides in
  // echarts-plotting.js).
  applyOptionOverrides(option) {
    if (!option.tooltip) return;
    let xsuffix = this.layout?.xaxis?.ticksuffix ?? "";
    let ysuffix = this.layout?.yaxis?.ticksuffix ?? "";
    // Matches the Plotly reference's native hovertemplate, which applies
    // `layout.xaxis.hoverformat`/`layout.yaxis.hoverformat` to `%{x}`/`%{y}` respectively.
    // NOTE: read off the active dataset's raw layout dict, not `this.layout` (see
    // bar.js's matching comment): the neutral `AxisIR` behind `this.layout` drops
    // `hoverformat`, a Plotly-only key.
    let dsLayout = this.datasets[this.activeDatasetIdx]?.layout;
    let xFormat = dsLayout?.xaxis?.hoverformat;
    let yFormat = dsLayout?.yaxis?.hoverformat;
    option.tooltip.formatter = (params) => {
      let [x, y] = params.value;
      return `<b>${params.name}</b><br/>X: ${window.formatWithHoverformat(x, xFormat)}${xsuffix}<br/>Y: ${window.formatWithHoverformat(y, yFormat)}${ysuffix}`;
    };
  }
}

// Make EchartsScatterPlot globally available (referenced by bare window.EchartsScatterPlot
// in echarts-plotting.js's initPlot(), not a bare identifier, so it's robust across the
// minified bundle regardless of module ordering after bundling).
window.EchartsScatterPlot = EchartsScatterPlot;
